airbyte-integrations/connectors/source-netsuite-odbc/source_netsuite_odbc/streams.py

```diff
@@ -0,0 +1,424 @@
+        if filter_string != "":
+          filter_string = filter_string + " OR " + filter_string_for_step
+        else:
+          filter_string = filter_string_for_step
+      return f"({filter_string})"
+
+
+    def generate_primary_key_filter_deprecated(self):
+      primary_key_filter = ""
+      # primary keys can be composite!  To deal with this, we add filters for each column
+      # This is an edge case.  In the standard case where there's only a single primary key column, 
+      # this could should only add a single filter
+      for key in self.primary_key_column:
+        last_value_seen = self.primary_key_last_value_seen[key] if key in self.primary_key_last_value_seen else STARTING_PRIMARY_KEY_VALUE
+        if primary_key_filter != "":
+          primary_key_filter = primary_key_filter + " AND "
+        if primary_key_filter == "" and self.has_composite_primary_key():
+          primary_key_filter = primary_key_filter + f"{key} = {last_value_seen}"
+        else:
+          primary_key_filter = primary_key_filter + f"{key} > {last_value_seen}"
+      # if it's a composite key, we add a second column to try to "upgrade" to the next column value
+      # if we don't do this, we don't properly make our way up the table
+      # NOTE:  This only works when there are two columns in the primary key
+      if self.has_composite_primary_key():
+        first_column = self.primary_key_column[0]
+        first_column_value = self.primary_key_last_value_seen[first_column] if first_column in self.primary_key_last_value_seen else STARTING_PRIMARY_KEY_VALUE
+        primary_key_filter = primary_key_filter + f" OR {first_column} > {first_column_value}"
+      return primary_key_filter
+    
+    def generate_primary_key_sorter(self):
+      primary_key_sorter = ""
+      for key in self.primary_key_column:
+        if primary_key_sorter != "":
+          primary_key_sorter = primary_key_sorter + ", "
+        primary_key_sorter = primary_key_sorter + f"{key} ASC"
+      return primary_key_sorter
+    
+    # How does Incremental Refresh work?
+    # We first split things into stream slices.  Then, we start with filtering our SQL query by > the start date of the stream slice
+    # and <= the end date of the stream slice.  We then sort by the incremental column, and then by the primary key column (s).
+    #  As we fetch records, we start filtering by > the most recent incremental time we've seen. We also add on a filter for 
+    # = the most recent incremental time but with a primary key > the last one we've seen.  This ensures that we don't miss any records
+    # in the case where page size = N and there are N + 1 records with the same incremental time. 
+    def generate_incremental_query(self, stream_slice):
+      if self.incremental_column is None:
+        raise AirbyteTracedException.from_exception(message=GENERATING_INCREMENTAL_QUERY_WITH_NO_INCREMENTAL_COLUMN_ERROR, failure_type=FailureType.system_error)
```

Comment:
from_exception expects an exception as parameter. You can instead directly create the traced exception
`AirbyteTracedException(message=GENERATING_INCREMENTAL_QUERY_WITH_NO_INCREMENTAL_COLUMN_ERROR, failure_type=FailureType.system_error)`